python/ray/tune/utils/file_transfer.py

```diff
@@ -217,9 +217,14 @@ def _get_recursive_files_and_stats(path: str) -> Dict[str, Tuple[float, int]]:
     for root, dirs, files in os.walk(path, topdown=False):
         rel_root = os.path.relpath(root, path)
         for file in files:
-            key = os.path.join(rel_root, file)
-            stat = os.lstat(os.path.join(path, key))
-            files_stats[key] = stat.st_mtime, stat.st_size
+            try:
+                key = os.path.join(rel_root, file)
+                stat = os.lstat(os.path.join(path, key))
+                files_stats[key] = stat.st_mtime, stat.st_size
+            except FileNotFoundError:
```

Comment:
I'd prefer just having this explicit error here and not a catch-all so we find other regressions, any objections?

Reply:
do we understand the root cause here?
I am still quite confused tbh. race condition between who and who?
Also should we log a warning or something for debugging purposes?

Reply:
The root cause is:

- Function trainable is asked to save_to_object() but has no checkpoint stored
- Function trainable thus creates null checkpoint directory
- PBT decides to PAUSE trial
- TrialRunner send stop to trial
- This calls trainable.stop.remote(), which calls trainable.cleanup(), which deletes the temporary checkpoint dir
- Additionally, trial runner schedules a `sync_dir_between_nodes()` to sync the trial directory on the node to the driver
- This happens in a different task though, so the cleanup() and the packing run into a race condition: packing still `os.walks()` through the directory, but cleanup() deletes some files while it is doing so
- This leads to the FileNotFoundError

I don't believe we should log these messages, it will clutter the output and the cause is quite clear and valid.
Instead, we should not sync temporary directories to begin with, see https://github.com/ray-project/ray/pull/27174

Reply:
what you described applies to temporary checkpoint dir, but I thought the problem is with the null checkpoint (checkpoint_-00001)? 
If I am tracing the code correctly, I can't find where it's deleted? So I'm confused as to where the racing is introduced.

Reply:
It's here: https://github.com/ray-project/ray/blob/master/python/ray/tune/trainable/function_trainable.py#L559

```
        if self.temp_checkpoint_dir is not None and os.path.exists(
            self.temp_checkpoint_dir
        ):
            shutil.rmtree(self.temp_checkpoint_dir)
            logger.debug("Clearing temporary checkpoint: %s", self.temp_checkpoint_dir)
```

And it's previously set here:

```
    def restore_from_object(self, obj):
        self.temp_checkpoint_dir = FuncCheckpointUtil.mk_temp_checkpoint_dir(
            self.logdir
        )
```

Reply:
Ah sorry, this is the temp checkpoint dir. Let me take another look